### PR TITLE
Add a facility to create cookies from a guest conversion response

### PR DESCRIPTION
**Pull request by ostapneko**
Creates GU cookies from the response that we get from registering a guest user.

See https://github.com/guardian/subscriptions-frontend/pull/236

### CODE REVIEW

File: src/test/scala/com/gu/identity/play/CookieBuilderTest.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package com.gu.identity.play
+
+import org.joda.time.format.ISODateTimeFormat
+import org.joda.time.{DateTime, Duration}
+import org.scalactic.Tolerance._
```
Comment (rtyley) [inline src/test/scala/com/gu/identity/play/CookieBuilderTest.scala:5]: I'm curious, this looks like the only use of scalactic, but where is it actually getting used? Do we actually need the dependency?

Comment (ostapneko) [inline src/test/scala/com/gu/identity/play/CookieBuilderTest.scala:5]: Used there https://github.com/guardian/identity-play-auth/pull/3/files#diff-3d48a75af0e0879fd059283dcba52731R56
But never heard of scalatestplus, so maybe it does it as well ?

Comment (rtyley) [inline src/test/scala/com/gu/identity/play/CookieBuilderTest.scala:5]: Ah, ok - so it's adding a bit of tolerance to cope with race conditions - I can live with that.

File: src/test/scala/com/gu/identity/play/CookieBuilderTest.scala [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      val domain = Some("domain")
+      val idCookies = CookieBuilder.fromGuestConversion(json, domain)
+      val Some(Seq(guuCookie, scguuCookie, scgulaCookie)) = idCookies
+
+      assert(guuCookie.name === "GU_U")
+      assert(guuCookie.value === "gu_u_value")
+      assert(guuCookie.maxAge.get === (90 * 24 * 60 * 60) +- 5)
+      assert(!guuCookie.secure)
+      assert(!guuCookie.httpOnly)
+      assert(guuCookie === domain)
```
Comment (rtyley) [inline src/test/scala/com/gu/identity/play/CookieBuilderTest.scala]: This is a bit odd, but I guess it passes!? Looks like your saying that a `Cookie` case class equals `Some("domain")` ?

File: src/main/scala/com/gu/identity/play/CookieBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      )(Cookie.apply(_, _))
+
+    for {
+      expirationString <- (json \ "cookies" \ "expiresAt").asOpt[String]
+      expiration <- Try { new DateTime(expirationString) }.toOption
+      maxAge = new Duration(DateTime.now, expiration).getStandardSeconds.toInt
+      cookies <- (json \ "cookies" \ "values").asOpt[Seq[Cookie]]
+      guuCookie <- cookies.find(_.name == "GU_U")
+      scguuCookie <- cookies.find(_.name == "SC_GU_U")
+      scgulaCookie <- cookies.find(_.name == "SC_GU_LA")
```
Comment (rtyley) [inline src/main/scala/com/gu/identity/play/CookieBuilder.scala]: The code above is constrained to the 3 known cookies, and I'd prefer that we match the [Frontend approach](https://github.com/guardian/frontend/blob/b17f08a/identity/app/services/PlaySigninService.scala#L19-L24) which honours _whatever_ list of cookies the Identity API response wants us to drop (tho' I agree that ideally the information about `httpOnly` and `secure` would be included in the API response, to reduce the amount of magic done by respecting `SC_`).

File: src/test/scala/com/gu/identity/play/CookieBuilderTest.scala (hunk 2 above)
Comment (ostapneko) [inline src/test/scala/com/gu/identity/play/CookieBuilderTest.scala]: This is a silly mistake I made, and no idea at all why it passes

File: src/main/scala/com/gu/identity/play/CookieBuilder.scala (hunk 3 above)
Comment (ostapneko) [inline src/main/scala/com/gu/identity/play/CookieBuilder.scala]: Unfortunately we have different rules for each of them. Yes, we have this SC_ rule, but how do we decide which one has a max-age for instance ?

File: src/test/scala/com/gu/identity/play/CookieBuilderTest.scala (hunk 2 above)
Comment (ostapneko) [inline src/test/scala/com/gu/identity/play/CookieBuilderTest.scala]: In fact it doesn't pass

File: src/main/scala/com/gu/identity/play/CookieBuilder.scala (hunk 3 above)
Comment (rtyley) [inline src/main/scala/com/gu/identity/play/CookieBuilder.scala]: > Yes, we have this SC_ rule, but how do we decide which one has a max-age for instance ?

That particular bit of information is already returned by the Identity API with a `sessionCookie` flag:

https://github.com/guardian/identity/blob/bb4a7611e/identity-api/src/main/scala/com/gu/identity/api/services/CookieResponseFactory.scala#L37

https://github.com/guardian/frontend/blob/fd2614c1/identity/app/idapiclient/responses/CookiesResponse.scala#L5-L7

...so you can use that!

Comment (ostapneko) [inline src/main/scala/com/gu/identity/play/CookieBuilder.scala]: OK

**Comment by rtyley** [comment]
:+1:

**Pull request merged**